plugins/producer_plugin/producer_plugin.cpp

```diff
@@ -1966,19 +1884,31 @@ producer_plugin_impl::start_block_result producer_plugin_impl::start_block() {
          return start_block_result::waiting_for_block;
    }
 
-   _pending_block_deadline         = block_timing_util::calculate_block_deadline(_cpu_effort_us, _pending_block_mode, block_time);
-   auto     preprocess_deadline    = _pending_block_deadline;
-   uint32_t production_round_index = block_timestamp_type(block_time).slot % chain::config::producer_repetitions;
-   if (production_round_index == 0) {
-      // first block of our round, wait for block production window
-      const auto start_block_time = block_time.to_time_point() - fc::microseconds(config::block_interval_us);
-      if (now < start_block_time) {
-         fc_dlog(_log, "Not starting block until ${bt}", ("bt", start_block_time));
-         schedule_delayed_production_loop(weak_from_this(), start_block_time);
-         return start_block_result::waiting_for_production;
+   if (in_producing_mode()) {
+      uint32_t production_round_index = block_timestamp_type(block_time).slot % chain::config::producer_repetitions;
+      if (production_round_index == 0) {
+         // first block of our round, wait for block production window
+         const auto start_block_time = block_time.to_time_point() - fc::microseconds(config::block_interval_us);
+         if (now < start_block_time) {
+            fc_dlog(_log, "Not starting block until ${bt}", ("bt", start_block_time));
+            schedule_delayed_production_loop(weak_from_this(), start_block_time);
+            return start_block_result::waiting_for_production;
+         }
       }
+
+      _pending_block_deadline = block_timing_util::calculate_producing_block_deadline(_cpu_effort_us, block_time);
+   } else if (!_producers.empty()) {
+      // cpu effort percent doesn't matter for the first block of the round, use max (block_interval_us) for cpu effort
+      auto wake_time = block_timing_util::calculate_producer_wake_up_time(config::block_interval_us, chain.head_block_num(), chain.head_block_time(),
+                                                                          _producers, chain.head_block_state()->active_schedule.producers,
+                                                                          _producer_watermarks);
+      _pending_block_deadline = wake_time ? *wake_time : fc::time_point::maximum();
+   } else {
+      _pending_block_deadline = fc::time_point::maximum();
```

Comment:
I wonder if it should be:

` _pending_block_deadline = chain.head_block_time() + fc::seconds(5);`

To be consistent with line 1883:

```
   if (in_speculating_mode()) {
      auto head_block_age = now - chain.head_block_time();
      if (head_block_age > fc::seconds(5))
         return start_block_result::waiting_for_block;
   }
```

Reply:
Excellent point. I wrote up a paragraph why it shouldn't correspond to the other 5 seconds and in doing so convinced myself of the opposite. Thanks.

I do wonder the wisdom of the current 5 second limit on speculative block production. The idea is that you don't want to speculative execute on too old of state. However, if a block producer misses a round then this node will stop processing trxs and wait for a block. Since everyone is doing that, then all trxs are held up until a block producer produces a block and sends it along.

I wonder if the new IF provides us with a different liveliness we could use to help make better determination. If we know that a series of blocks were not produced we could safely speculative execute trxs on older state.

Reply:
If the node is connected to multiple peers, shouldn't we assume that its state will be up-to-date, as it will receive blocks when they are produced? So I don't see the rationale to assume that our state is old vs the whole chain (again provided that we are connected to multiple peers).

Reply:
See updated PR description.